rewrite-javascript/src/main/javascript/recipe.ts

```diff
@@ -0,0 +1,221 @@
+    }
+}
+
+export interface RecipeDescriptor {
+    readonly displayName: string
+    readonly instanceName: string
+    readonly description: string
+    readonly tags: string[]
+    readonly estimatedEffortPerOccurrence: Minutes
+    readonly options: ({ name: string } & OptionDescriptor)[]
+}
+
+export interface OptionDescriptor {
+    readonly displayName: string
+    readonly description: string
+    readonly example?: string
+    readonly valid?: string[],
+}
+
+export abstract class ScanningRecipe<P> extends Recipe {
+    private readonly recipeAccMessage = Symbol("org.openrewrite.recipe.acc");
+
+    accumulator(cursor: Cursor, ctx: ExecutionContext): P {
+        const ms = cursor.root.messages;
+        if (!ms.has(this.recipeAccMessage)) {
+            ms.set(this.recipeAccMessage, this.initialValue(ctx));
+        }
+        return ms.get(this.recipeAccMessage);
+    }
+
+    abstract initialValue(ctx: ExecutionContext): P
+
+    get editor(): TreeVisitor<any, ExecutionContext> {
+        const editorWithContext = (cursor: Cursor, ctx: ExecutionContext) =>
+            this.editorWithData(this.accumulator(cursor, ctx));
+
+        return new class extends TreeVisitor<any, ExecutionContext> {
+            private delegate?: TreeVisitor<any, ExecutionContext>
+
+            isAcceptable(sourceFile: SourceFile, ctx: ExecutionContext): boolean {
+                return this.delegateForCtx(ctx).isAcceptable(sourceFile, ctx);
+            }
+
+            async visit<R extends Tree>(tree: Tree, ctx: ExecutionContext, parent?: Cursor): Promise<R | undefined> {
+                return this.delegateForCtx(ctx).visit(tree, ctx, parent);
+            }
+
+            private delegateForCtx(ctx: ExecutionContext) {
+                if (!this.delegate) {
+                    this.delegate = editorWithContext(this.cursor, ctx);
+                }
+                return this.delegate;
+            }
+        }
+    }
+
+    editorWithData(acc: P): TreeVisitor<any, ExecutionContext> {
+        return noopVisitor();
+    }
+
+    generate(acc: P): SourceFile[] {
+        return [];
+    }
+
+    scanner(acc: P): TreeVisitor<any, ExecutionContext> {
+        return noopVisitor();
+    }
+}
+
+/**
+ * Do not permit overriding of editor()
+ */
+Object.freeze(ScanningRecipe.prototype.editor);
+
+export class RecipeRegistry {
+    // The registry map stores recipe constructors keyed by their instance name.
+    static all = new Map<string, { new(options?: {}): Recipe }>();
+
+    public static register(name: string, recipeClass: { new(options?: {}): Recipe }): void {
+        RecipeRegistry.all.set(name, recipeClass);
+    }
+}
+
+/**
+ * @param recipeName A unique name for the recipe consisting of a dot separated sequence of category
+ * names in which the recipe should appear followed by a name. For example,
+ * "org.openrewrite.typescript.find-methods-by-pattern".
+ */
+export function Registered(recipeName: string) {
```

Comment:
I like this

I could image a @Recipe decorator that accepts an object with recipe meta data and standardized across all languages supporting the decorator pattern to have the authorship experience converge slightly.  Would also reduce the code in the recipe down to just method implementations (javascript is already pretty slim but for other languages may reduce some boilerplate)

Reply:
examples:

typescript
```ts
@Recipe({
  id: 'org.testkitchen.BigAwesome', 
  name: 'Big awesome', 
  description: 'A recipe ensuring big awesome things'
})
class BigAwesome extends Recipe {}
```

java
```java
@Recipe(
    id = "org.testkitchen.BigAwesome",
    name = "Big awesome",
    description = "A recipe ensuring big awesome things"
)
public class BigAwesome extends Recipe {}
```

python
```python
@recipe(
    id="org.testkitchen.BigAwesome",
    name="Big awesome",
    description="A recipe ensuring big awesome things"
)
class BigAwesome(Recipe):
```

c#
```csharp
[Recipe(
    Id = "org.testkitchen.BigAwesome",
    Name = "Big awesome",
    Description = "A recipe ensuring big awesome things"
)]
public class BigAwesome : Recipe {}
```

Reply:
I don't think I'd like the decorator pattern here because there is no way to force compile time application of it.